bolt-sidecar/src/api/commitments/firewall_recv.rs

```diff
@@ -0,0 +1,548 @@
+            }
+        }
+
+        // Reconnect on failure
+        // TODO: add backoff
+        warn!(?url, "connection lost. reconnecting...");
+        tokio::time::sleep(Duration::from_secs(5)).await;
+    }
+}
+
+/// The [CommitmentRequestProcessor] handles incoming commitment requests a the websocket
+/// connection, and forwards them to the [CommitmentEvent] tx channel for processing.
+struct CommitmentRequestProcessor {
+    /// The URL of the connected websocket server.
+    url: Url,
+    /// The channel to send commitment events to be processed.
+    api_events_tx: mpsc::Sender<CommitmentEvent>,
+    /// The websocket writer sink.
+    write_sink: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+    /// The websocket reader stream.
+    read_stream: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+    /// The receiver for keep-alive ping messages.
+    ping_rx: broadcast::Receiver<()>,
+    /// The collection of pending commitment requests responses, sent with [api_events_tx].
+    /// NOTE: Is there a better way to avoid this monster type?
+    /// SAFETY: the `poll` implementation of this struct promptly handles these responses and
+    /// ensures this vector doesn't grow indefinitely.
+    pending_commitment_responses: FuturesUnordered<
+        Pin<
+            Box<
+                dyn Future<
+                        Output = Result<
+                            Result<SignedCommitment, CommitmentError>,
+                            oneshot::error::RecvError,
+                        >,
+                    > + Send,
+            >,
+        >,
+    >,
+    /// The collection of outgoing messages to be sent to the connected websocket server.
+    outgoing_messages: VecDeque<Message>,
+}
+
+impl CommitmentRequestProcessor {
+    pub fn new(
+        url: Url,
+        tx: mpsc::Sender<CommitmentEvent>,
+        write: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+        read: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+        ping_rx: broadcast::Receiver<()>,
+    ) -> Self {
+        Self {
+            url,
+            api_events_tx: tx,
+            write_sink: write,
+            read_stream: read,
+            ping_rx,
+            pending_commitment_responses: FuturesUnordered::new(),
+            outgoing_messages: VecDeque::new(),
+        }
+    }
+}
+
+impl Future for CommitmentRequestProcessor {
+    type Output = ();
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        let rpc_url = this.url.clone();
+
+        loop {
+            let mut progress = false;
+
+            // 1. Handle incoming websocket messages from the read stream.
+            while let Poll::Ready(maybe_message) = this.read_stream.poll_next_unpin(cx) {
+                progress = true;
+
+                match maybe_message {
```

Comment:
overall the matching going on here feels a bit heavy!
can you make the control flow more linear by using something like:

```rust
let Some(Ok(message)) = maybe_message else {
	warn!("websocket connection with {rpc_url} dropped");
	return Poll::Ready(());
};

match message {
	Message::Text(...
	Message::Close(_) ...
```